types/stampit/v3/index.d.ts

```diff
@@ -0,0 +1,525 @@
+// Type definitions for stampit 3.0
+// Project: https://github.com/stampit-org/stampit
+// Definitions by: Vasyl Boroviak <https://github.com/koresar>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/**
+ * Function used as .init() argument.
+ */
+interface Init {
+    (factoryArg: any, ctx?: Context): any;
+}
+
+/**
+ * Composer function
+ */
+interface Composer {
+    ({ stamp, composables }: { stamp: Stamp; composables: Composable[] }): any;
+}
+
+/** The stamp Descriptor */
+interface Descriptor {
+    /** Create a new stamp based on this descriptor */
+    (...composables: Composable[]): Stamp;
+    /**
+     * A hash containing methods (functions) of any future created instance.
+     */
+    methods?: {};
+    /**
+     * Initialization function(s) which will be called per each newly created
+     * instance.
+     */
+    initializers?: Init[];
+    /**
+     * Properties which will shallowly copied into any future created instance.
+     */
+    properties?: {};
+    /**
+     * Properties which will be mixed to the new and any other stamp which this stamp will be composed with.
+     */
+    staticProperties?: {};
+    /** Deeply merged properties of object instances */
+    deepProperties?: {};
+    /** ES5 Property Descriptors applied to object instances */
+    propertyDescriptors?: {};
+    /** Deeply merged properties of Stamps */
+    staticDeepProperties?: {};
+    /** ES5 Property Descriptors applied to Stamps */
+    staticPropertyDescriptors?: {};
+    /** A configuration object to be shallowly assigned to Stamps */
+    configuration?: {};
+    /** A configuration object to be deeply merged to Stamps */
+    deepConfiguration?: {};
+}
+
+/** Any composable object (stamp or descriptor) */
+type Composable = Stamp | Descriptor;
+
+/**
+ * The .init() function argument.
+ */
+interface Context {
+    /**
+     * The object which has been just instantiated.
+     */
+    instance: any;
+
+    /**
+     * The stamp the object has been instantiated with.
+     */
+    stamp: Stamp;
+
+    /**
+     * The arguments list passed to the stamp.
+     */
+    args: any[];
+}
+
+interface Options {
+    /**
+     * A hash containing methods (functions) of any future created instance.
+     */
+    methods?: {} | {}[];
```

Comment:
Most of these `{} | {}[]` are really equivalent to `{}`.

Reply:
I'll make the change, but why is that? I'm having trouble finding an explanation in the Typescript docs.

Reply:
It's because `{}` means "I take anything apart from `null` or `undefined` because anything satisfies that structure. So once you've said that, adding an array really provides no added value unless you're planning on coming back and amending the definitions.

Reply:
I see. Thanks for the explanation!